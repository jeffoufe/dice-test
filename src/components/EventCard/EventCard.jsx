// @flow

import React from 'react';
import styles from './EventCard.module.scss';
import PlayButton from '../PlayButton/PlayButton';
import Featured from '../Featured/Featured';
import OnSaleDate from '../OnSaleDate/OnSaleDate';
import MoreInfo from '../MoreInfo/MoreInfo';
import Button from '../Button/Button';
import { Text, ImportantText } from '../Typography/Typography';
import buildPrice from '../../utils/buildPrice';
import moment from 'moment';

import type { LineUp } from '../../models/LineUp';
import type { Ticket } from '../../models/Ticket';

export type EventCardProps = {
    event_images: {
        landscape: string,
        square: string
    },
    apple_music_tracks: Array<Object>,
    spotify_tracks: Array<Object>,
    ticket_types: Array<Ticket>,
    featured: boolean,
    sale_start_date: string,
    featured: boolean,
    date: string,
    name: string,
    venue: string,
    description: string,
    sold_out: boolean,
    url: string,
    lineup: LineUp,
    currency: string,
    location: {
        city: string,
        country: string
    }
}

const EventCard = (props: EventCardProps) => {
    const eventDate = moment(props.date);
    const isNotOnSaleYet = moment(props.sale_start_date).isAfter(moment());

    const sortedTicketTypes = props.ticket_types.sort((ticketA: Ticket, ticketB: Ticket) => ticketA.price.total - ticketB.price.total);
    
    return (
        <div className={styles['event-card-container']}>
            <div className={styles['event-card-image-container']}>
                <img src={props.event_images[props.featured ? 'landscape' : 'square']} className={styles['event-card-image']} alt={props.name} />
                {!!(props.apple_music_tracks.length || props.spotify_tracks.length) && <PlayButton />}
                {props.featured && <Featured />}
                {isNotOnSaleYet && <OnSaleDate date={props['sale_start_date']} />}
            </div>

            <Text className={styles['event-card-date']}>
                {eventDate.format('ddd D MMM')} 
                &#65293;
                {eventDate.format('LT')}
            </Text>

            <div className={styles['event-card-general-info']}>
                <h2 className={styles['event-card-name']}>
                    <a href={props.url} target='_blank' rel="noopener noreferrer">{props.name}</a>
                </h2>
                <ImportantText>{props.venue}</ImportantText>
                <Text>{`${props.location.city}, ${props.location.country}`}</Text>
            </div>

            <MoreInfo {...props} />

            <div className={styles['event-card-actions']}>
                <Button theme={props['sold_out'] ? 'disabled' : 'blue'}>
                    {props['sold_out'] 
                        ? 'Sold out'
                        : isNotOnSaleYet 
                            ? 'Get reminded' 
                            : 'Book now'
                    }
                </Button>
                <div className={styles['event-card-price']}>
                    {props.ticket_types.length > 1 && (
                        <div className={styles['event-card-price-from']}>
                           From
                        </div>
                    )}
                    <div className={styles['event-card-price-value']}>{buildPrice(props.currency, sortedTicketTypes[0].price.total)}</div>
                </div>
            </div>
        </div>
    );
}

export default EventCard;