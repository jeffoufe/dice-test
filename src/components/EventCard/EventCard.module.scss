@import '../../constants/_colors';
@import '../../constants/_breakpoints';

.container {
    display: inline-block;
    width: 100%;
    margin-bottom: 30px;

    @media screen and (min-width: $tablet) {
        width: 320px;
        margin: 32px 24px;
    }
}

.image-container {
    position: relative;
}

@import '../../constants/breakpoints';

.image {
    width: 100%;
}

.date {
    margin-top: 8px;
}

.general {
    @media screen and (min-width: $tablet) {
        height: 158px;
    }
}

.name {
    font-size: 28px;
    line-height: 32px;
    margin-top: 10px;
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;

    > a {
        color: $black;
        text-decoration: none;
    }
}

.actions {
    margin-top: 25px;
    display: flex;
    justify-content: space-between;
}

.price {
    &-from {
        color: $black;
        opacity: 0.5;
        margin-bottom: 5px;

        + .price-value {
            line-height: unset;
        }
    }

    &-value {
        color: $black;
        font-size: 32px;
        line-height: 40px;
    }
}