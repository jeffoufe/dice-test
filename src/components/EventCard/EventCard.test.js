import EventCard from './EventCard';
import { shallow } from 'enzyme';
import React from 'react';

describe('EventCard', () => {
    const props = {
        event_images: {
            landscape: 'landscape',
            square: 'square'
        },
        apple_music_tracks: [],
        spotify_tracks: [],
        ticket_types: [{ price: { total: 800 } }, { price: { total: 400 } }],
        featured: false,
        sale_start_date: '2019-08-19',
        date: 'date',
        name: 'name',
        venue: 'venue',
        description: 'description',
        sold_out: false,
        url: 'url',
        lineup: [],
        currency: 'GBP',
        location: {
            city: 'city',
            country: 'country'
        }
    }

    it('Not sold out, not featured', () => {
        const wrapper = shallow(<EventCard {...props} />);
        expect(wrapper).toMatchSnapshot();
    })

    it('sold out, not featured', () => {
        const newProps = { ...props, sold_out: true }
        const wrapper = shallow(<EventCard {...newProps} />);
        expect(wrapper).toMatchSnapshot();
    })

    it('not sold out, featured', () => {
        const newProps = { ...props, featured: true }
        const wrapper = shallow(<EventCard {...newProps} />);
        expect(wrapper).toMatchSnapshot();
    })

    it('not on sale yet', () => {
        const newProps = { ...props, sale_start_date: '2089-08-19' };
        const wrapper = shallow(<EventCard {...newProps} />);
        expect(wrapper).toMatchSnapshot();
    })
})