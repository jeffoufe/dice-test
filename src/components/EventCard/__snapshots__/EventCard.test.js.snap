// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard Not sold out, not featured 1`] = `
<div
  className="event-card-container"
>
  <div
    className="event-card-image-container"
  >
    <img
      alt="name"
      className="event-card-image"
      src="square"
    />
  </div>
  <Text
    className="event-card-date"
  >
    Invalid date
    －
    Invalid date
  </Text>
  <div
    className="event-card-general-info"
  >
    <h2
      className="event-card-name"
    >
      <a
        href="url"
        rel="noopener noreferrer"
        target="_blank"
      >
        name
      </a>
    </h2>
    <ImportantText>
      venue
    </ImportantText>
    <Text>
      city, country
    </Text>
  </div>
  <MoreInfo
    apple_music_tracks={Array []}
    currency="GBP"
    date="date"
    description="description"
    event_images={
      Object {
        "landscape": "landscape",
        "square": "square",
      }
    }
    featured={false}
    lineup={Array []}
    location={
      Object {
        "city": "city",
        "country": "country",
      }
    }
    name="name"
    sale_start_date="2019-08-19"
    sold_out={false}
    spotify_tracks={Array []}
    ticket_types={
      Array [
        Object {
          "price": Object {
            "total": 400,
          },
        },
        Object {
          "price": Object {
            "total": 800,
          },
        },
      ]
    }
    url="url"
    venue="venue"
  />
  <div
    className="event-card-actions"
  >
    <Button
      theme="blue"
    >
      Book now
    </Button>
    <div
      className="event-card-price"
    >
      <div
        className="event-card-price-from"
      >
        From
      </div>
      <div
        className="event-card-price-value"
      >
        £4
      </div>
    </div>
  </div>
</div>
`;

exports[`EventCard not on sale yet 1`] = `
<div
  className="event-card-container"
>
  <div
    className="event-card-image-container"
  >
    <img
      alt="name"
      className="event-card-image"
      src="square"
    />
    <OnSaleDate
      date="2089-08-19"
    />
  </div>
  <Text
    className="event-card-date"
  >
    Invalid date
    －
    Invalid date
  </Text>
  <div
    className="event-card-general-info"
  >
    <h2
      className="event-card-name"
    >
      <a
        href="url"
        rel="noopener noreferrer"
        target="_blank"
      >
        name
      </a>
    </h2>
    <ImportantText>
      venue
    </ImportantText>
    <Text>
      city, country
    </Text>
  </div>
  <MoreInfo
    apple_music_tracks={Array []}
    currency="GBP"
    date="date"
    description="description"
    event_images={
      Object {
        "landscape": "landscape",
        "square": "square",
      }
    }
    featured={false}
    lineup={Array []}
    location={
      Object {
        "city": "city",
        "country": "country",
      }
    }
    name="name"
    sale_start_date="2089-08-19"
    sold_out={false}
    spotify_tracks={Array []}
    ticket_types={
      Array [
        Object {
          "price": Object {
            "total": 400,
          },
        },
        Object {
          "price": Object {
            "total": 800,
          },
        },
      ]
    }
    url="url"
    venue="venue"
  />
  <div
    className="event-card-actions"
  >
    <Button
      theme="blue"
    >
      Get reminded
    </Button>
    <div
      className="event-card-price"
    >
      <div
        className="event-card-price-from"
      >
        From
      </div>
      <div
        className="event-card-price-value"
      >
        £4
      </div>
    </div>
  </div>
</div>
`;

exports[`EventCard not sold out, featured 1`] = `
<div
  className="event-card-container"
>
  <div
    className="event-card-image-container"
  >
    <img
      alt="name"
      className="event-card-image"
      src="landscape"
    />
    <Featured />
  </div>
  <Text
    className="event-card-date"
  >
    Invalid date
    －
    Invalid date
  </Text>
  <div
    className="event-card-general-info"
  >
    <h2
      className="event-card-name"
    >
      <a
        href="url"
        rel="noopener noreferrer"
        target="_blank"
      >
        name
      </a>
    </h2>
    <ImportantText>
      venue
    </ImportantText>
    <Text>
      city, country
    </Text>
  </div>
  <MoreInfo
    apple_music_tracks={Array []}
    currency="GBP"
    date="date"
    description="description"
    event_images={
      Object {
        "landscape": "landscape",
        "square": "square",
      }
    }
    featured={true}
    lineup={Array []}
    location={
      Object {
        "city": "city",
        "country": "country",
      }
    }
    name="name"
    sale_start_date="2019-08-19"
    sold_out={false}
    spotify_tracks={Array []}
    ticket_types={
      Array [
        Object {
          "price": Object {
            "total": 400,
          },
        },
        Object {
          "price": Object {
            "total": 800,
          },
        },
      ]
    }
    url="url"
    venue="venue"
  />
  <div
    className="event-card-actions"
  >
    <Button
      theme="blue"
    >
      Book now
    </Button>
    <div
      className="event-card-price"
    >
      <div
        className="event-card-price-from"
      >
        From
      </div>
      <div
        className="event-card-price-value"
      >
        £4
      </div>
    </div>
  </div>
</div>
`;

exports[`EventCard sold out, not featured 1`] = `
<div
  className="event-card-container"
>
  <div
    className="event-card-image-container"
  >
    <img
      alt="name"
      className="event-card-image"
      src="square"
    />
  </div>
  <Text
    className="event-card-date"
  >
    Invalid date
    －
    Invalid date
  </Text>
  <div
    className="event-card-general-info"
  >
    <h2
      className="event-card-name"
    >
      <a
        href="url"
        rel="noopener noreferrer"
        target="_blank"
      >
        name
      </a>
    </h2>
    <ImportantText>
      venue
    </ImportantText>
    <Text>
      city, country
    </Text>
  </div>
  <MoreInfo
    apple_music_tracks={Array []}
    currency="GBP"
    date="date"
    description="description"
    event_images={
      Object {
        "landscape": "landscape",
        "square": "square",
      }
    }
    featured={false}
    lineup={Array []}
    location={
      Object {
        "city": "city",
        "country": "country",
      }
    }
    name="name"
    sale_start_date="2019-08-19"
    sold_out={true}
    spotify_tracks={Array []}
    ticket_types={
      Array [
        Object {
          "price": Object {
            "total": 400,
          },
        },
        Object {
          "price": Object {
            "total": 800,
          },
        },
      ]
    }
    url="url"
    venue="venue"
  />
  <div
    className="event-card-actions"
  >
    <Button
      theme="disabled"
    >
      Sold out
    </Button>
    <div
      className="event-card-price"
    >
      <div
        className="event-card-price-from"
      >
        From
      </div>
      <div
        className="event-card-price-value"
      >
        £4
      </div>
    </div>
  </div>
</div>
`;
